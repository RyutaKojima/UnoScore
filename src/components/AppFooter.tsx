import React from 'react'
import clsx from 'clsx'

type Props = {
  className?: string
}

export const AppFooter: React.FC<Props> = ({ className }) => {
  return (
    <footer className={clsx('p-4 border-t', className)}>
      <p className="text-center">Powered by Ryuta Kojima</p>
    </footer>
  )
}
