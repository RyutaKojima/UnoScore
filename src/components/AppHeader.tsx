import React from 'react'
import clsx from 'clsx'

type Props = {
  className?: string
}

export const AppHeader: React.FC<Props> = ({ className }) => {
  return (
    <header className={clsx('p-4 border-b', className)}>
      <h1 className="text-center">UNO Score</h1>
    </header>
  )
}
