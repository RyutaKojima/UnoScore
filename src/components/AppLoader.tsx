import React from 'react'
import { Spinner } from './Spinner'

type Props = {
  loading: boolean
}

export const AppLoader: React.VFC<Props> = ({ loading }) => {
  return loading ? (
    <div className="fixed inset-0 flex justify-center items-center bg-white bg-opacity-50 z-10">
      <Spinner />
    </div>
  ) : null
}
