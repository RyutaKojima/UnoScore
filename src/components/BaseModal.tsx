import React, { useEffect, useRef } from 'react'
import ReactDOM from 'react-dom'
import clsx from 'clsx'
import {
  disableBodyScroll,
  enableBodyScroll,
  clearAllBodyScrollLocks,
} from 'body-scroll-lock'

type Props = {
  isShow: boolean
  className?: string
}

export const BaseModal: React.FC<Props> = ({ isShow, className, children }) => {
  const modalRef = useRef<HTMLDivElement>(null)

  useEffect(() => {
    if (!modalRef.current) return

    if (isShow) {
      disableBodyScroll(modalRef.current)
    } else {
      enableBodyScroll(modalRef.current)
    }

    return () => {
      clearAllBodyScrollLocks()
    }
  }, [isShow])

  const container = process.browser ? document.querySelector('#__next') : null

  return container && isShow
    ? ReactDOM.createPortal(
        <div className="fixed inset-0 bg-black bg-opacity-50 flex justify-center items-center p-4">
          <div
            ref={modalRef}
            className={clsx(
              'bg-white p-4 max-h-90vh w-full max-w-lg rounded-lg relative overflow-y-scroll',
              className
            )}
          >
            {children}
          </div>
          <style jsx>{`
            .max-h-90vh {
              max-height: 90vh;
            }
          `}</style>
        </div>,
        container
      )
    : null
}
