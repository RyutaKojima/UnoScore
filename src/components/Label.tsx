import React from 'react'
import clsx from 'clsx'

type Props = {
  className?: string
}

export const Label: React.FC<Props> = ({ className, children }) => {
  return (
    <span className={clsx('text-gray-700 text-sm font-semibold', className)}>
      {children}
    </span>
  )
}
