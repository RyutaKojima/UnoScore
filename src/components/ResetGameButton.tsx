import React from 'react'
import clsx from 'clsx'

type Props = {
  label?: string
  className?: string
  onReset: () => void
}

export const ResetGameButton: React.FC<Props> = ({
  label = 'リセット',
  className,
  onReset,
}) => {
  const handleOnClick = async () => {
    const result = confirm('ゲームをリセットしますか？')
    if (result) {
      onReset()
    }
  }

  return (
    <button
      className={clsx('btn bg-gray-800 text-white', className)}
      onClick={handleOnClick}
    >
      {label}
    </button>
  )
}
