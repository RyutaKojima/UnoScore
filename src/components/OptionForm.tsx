import React from 'react'
import { Options } from '../pages'
import { Label } from './Label'

type Props = {
  options: Options
  onChange: (newOptions: Options) => void
}

export const OptionForm = (props: Props): JSX.Element => (
  <div className="space-x-3">
    <label className="inline-flex items-center">
      <Label className="mr-2">２位免除</Label>
      <input
        type="checkbox"
        checked={props.options.rescueSecond}
        onChange={() => {
          props.onChange({
            ...props.options,
            rescueSecond: !props.options.rescueSecond,
          })
        }}
        className="form-checkbox"
      />
    </label>

    <label className="inline-flex items-center">
      <Label className="mr-2">３位免除</Label>
      <input
        type="checkbox"
        checked={props.options.rescueThird}
        onChange={() => {
          props.onChange({
            ...props.options,
            rescueThird: !props.options.rescueThird,
          })
        }}
        className="form-checkbox"
      />
    </label>
  </div>
)
