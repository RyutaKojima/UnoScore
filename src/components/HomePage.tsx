import React, { ChangeEvent, ComponentProps, useState } from 'react'
import { lastOfArray } from '../utils/utils'
import { ScoreTable } from './ScoreTable'
import { BaseLayout } from '../layouts/BaseLayout'
import { Section } from './Section'
import { OptionForm } from './OptionForm'
import { IOption } from '../interfaces/option'
import { Label } from './Label'
import { UserAppendForm } from './UserAppendForm'
import { Result } from './Result'
import { ResetGameButton } from './ResetGameButton'
import { IRound } from '../interfaces/round'
import { ExclamationIcon } from './icons/ExclamationIcon'
import clsx from 'clsx'

type Props = {
  rounds: IRound[]
  addRound: (numberOfPlayers: number) => void
  setScore: (roundIndex: number, playerIndex: number, score: number) => void
  players: string[]
  setPlayers: (players: string[]) => void
  option: IOption
  setOption: (option: IOption) => void
  initializeDatabase: () => void
}

export const HomePage: React.FC<Props> = ({
  rounds,
  addRound,
  setScore,
  players,
  setPlayers,
  option,
  setOption,
  initializeDatabase,
}) => {
  const [isForceChange, setIsForceChange] = useState(false)
  const [errors, setErrors] = useState<string[]>([])

  const isInputScoreValid = (): boolean => {
    const currentRound: null | number[] = lastOfArray<number[]>(rounds)
    if (currentRound === null) {
      return true
    }

    return currentRound.filter((score) => score === 0).length === 1
  }

  const handleOnChangeMagnification = (
    e: ChangeEvent<HTMLSelectElement>
  ): void => {
    setOption({ ...option, magnification: Number(e.target.value) })
  }

  const handleAppendUser = (name: string): void => {
    if (!name) {
      return
    }
    setPlayers([...players, name])
  }

  const handleOnChangeScore: ComponentProps<typeof ScoreTable>['onChange'] = (
    score,
    roundIndex,
    index
  ) => {
    setScore(roundIndex, index, score)
  }

  const handleAddRound = (): void => {
    if (players.length <= 1) {
      setErrors(['プレイヤーを登録してください'])
      return
    }
    if (!isInputScoreValid()) {
      setErrors(['スコアが未入力のプレイヤーがいます'])
      return
    }
    setErrors([])
    addRound(players.length)
  }

  const toggleForceChange = (): void => {
    if (!process.browser) {
      return
    }

    if (isForceChange) {
      setIsForceChange(false)
    } else {
      const result = window.confirm('強制修正モードを有効化しますか？')
      if (result) {
        setIsForceChange(true)
      }
    }
  }

  return (
    <BaseLayout
      containerClass="bg-gray-200"
      className="w-full max-w-xl mx-auto flex flex-col space-y-10 py-10 px-3"
    >
      <Section title="Step.1 Initial settings" className="space-y-4">
        <OptionForm
          options={option}
          onChange={(newOptions: IOption) => setOption(newOptions)}
        />

        <label className="block mt-4">
          <Label>倍率：</Label>
          <select
            value={option.magnification}
            className="form-select mt-1 block w-full"
            onChange={handleOnChangeMagnification}
          >
            <option value="1">x1</option>
            <option value="2">x2</option>
            <option value="3">x3</option>
            <option value="4">x4</option>
            <option value="5">x5</option>
          </select>
        </label>

        <UserAppendForm onAppended={handleAppendUser} />
      </Section>

      <Section title="Step.2 Input scores" className="relative">
        <div className="absolute right-0 top-0 p-4">
          <button className="flex items-center" onClick={toggleForceChange}>
            <ExclamationIcon
              className={clsx('w-6 h-6', {
                'text-gray-400': !isForceChange,
                'text-red-600': isForceChange,
              })}
            />
          </button>
        </div>
        <ScoreTable
          isForceChange={isForceChange}
          players={players}
          rounds={rounds}
          onChange={handleOnChangeScore}
        />
        <p className="text-red-400">{errors}</p>
        <div className="text-center mt-4">
          <button
            onClick={handleAddRound}
            className="btn bg-gradient-to-r from-teal-400 to-blue-500 text-white"
          >
            add Round
          </button>
        </div>
      </Section>

      <Section title="Step.3 Results">
        <Result players={players} rounds={rounds} options={option} />
      </Section>

      <Section title="Step.4 Clear">
        <div className="text-center mt-4">
          <ResetGameButton onReset={initializeDatabase} />
        </div>
      </Section>
    </BaseLayout>
  )
}
