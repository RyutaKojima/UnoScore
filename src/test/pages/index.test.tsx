import React from 'react'
import { render } from '../testUtils'
import { HomePage } from '../../components/HomePage'
// import { render, fireEvent } from '../testUtils'

describe('Home page', () => {
  it('matches snapshot', () => {
    const handleSet = () => {
      return
    }

    const { asFragment } = render(
      <HomePage
        loading={false}
        rounds={[]}
        players={[]}
        option={{
          rescueSecond: true,
          rescueThird: false,
          magnification: 1,
        }}
        addRound={handleSet}
        setScore={handleSet}
        setPlayers={handleSet}
        setOption={handleSet}
        initializeDatabase={handleSet}
      />,
      {}
    )
    expect(asFragment()).toMatchSnapshot()
  })
  //   it('clicking button triggers alert', () => {
  //   const { getByText } = render(<Home />, {})
  //   window.alert = jest.fn()
  //   fireEvent.click(getByText('Test Button'))
  //   expect(window.alert).toHaveBeenCalledWith('With typescript and Jest')
  //   })
})
