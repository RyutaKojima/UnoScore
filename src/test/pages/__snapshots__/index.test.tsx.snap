// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="min-h-screen flex flex-col bg-gray-200"
  >
    <header
      class="p-4"
    >
      <div
        class="flex justify-center items-center space-x-4"
      >
        <div
          style="max-width: 100%; width: 25px;"
        >
          <div
            style="position: relative; padding-bottom: 160%;"
          >
            <img
              class="shadow __lazy"
              data-src="/_next/image?url=%2Fimg%2Fdraw4.png&w=320&q=75"
              data-srcset="/_next/image?url=%2Fimg%2Fdraw4.png&w=320&q=75 320w"
              style="visibility: hidden; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
            />
          </div>
        </div>
        <h1
          class="text-2xl font-bold italic text-gray-800"
        >
          UNO Score
        </h1>
      </div>
    </header>
    <main
      class="flex-1 w-full max-w-xl mx-auto flex flex-col space-y-10 py-10 px-3"
    >
      <section
        class="bg-white p-4 rounded-lg shadow-xl space-y-4"
      >
        <h2
          class="text-2xl font-bold text-gray-800 mb-4"
        >
          Step.1 Initial settings
        </h2>
        <div
          class="space-x-3"
        >
          <label
            class="inline-flex items-center"
          >
            <span
              class="text-gray-700 text-sm font-semibold mr-2"
            >
              ２位免除
            </span>
            <input
              checked=""
              class="form-checkbox"
              type="checkbox"
            />
          </label>
          <label
            class="inline-flex items-center"
          >
            <span
              class="text-gray-700 text-sm font-semibold mr-2"
            >
              ３位免除
            </span>
            <input
              class="form-checkbox"
              type="checkbox"
            />
          </label>
        </div>
        <label
          class="block mt-4"
        >
          <span
            class="text-gray-700 text-sm font-semibold"
          >
            倍率：
          </span>
          <select
            class="form-select mt-1 block w-full"
          >
            <option
              value="1"
            >
              x1
            </option>
            <option
              value="2"
            >
              x2
            </option>
            <option
              value="3"
            >
              x3
            </option>
            <option
              value="4"
            >
              x4
            </option>
            <option
              value="5"
            >
              x5
            </option>
          </select>
        </label>
        <form>
          <label
            class="block"
          >
            <span
              class="text-gray-700 text-sm font-semibold"
            >
              ユーザー追加
            </span>
            <div
              class="flex mt-1"
            >
              <input
                class="form-input block w-full"
                type="text"
                value=""
              />
              <button
                class="btn bg-gradient-to-r from-teal-400 to-blue-500 text-white ml-4"
                type="submit"
              >
                add
              </button>
            </div>
          </label>
        </form>
      </section>
      <section
        class="bg-white p-4 rounded-lg shadow-xl relative"
      >
        <h2
          class="text-2xl font-bold text-gray-800 mb-4"
        >
          Step.2 Input scores
        </h2>
        <div
          class="absolute right-0 top-0 p-4"
        >
          <button
            class="flex items-center"
          >
            <svg
              class="w-6 h-6 text-gray-400"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 9v2m0 4h.01m-6.938 4h13.856c1.54 0 2.502-1.667 1.732-3L13.732 4c-.77-1.333-2.694-1.333-3.464 0L3.34 16c-.77 1.333.192 3 1.732 3z"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
        <div
          class="text-right"
        >
          <div>
            <div
              class="relative inline-block w-10 mr-2 align-middle select-none transition duration-200 ease-in"
            >
              <label>
                <input
                  checked=""
                  class="absolute block w-6 h-6 rounded-full bg-white border-4 transition duration-200 ease-in appearance-none cursor-pointer focus:outline-none right-0 border-teal-400"
                  type="checkbox"
                />
                <span
                  class="block overflow-hidden h-6 rounded-full cursor-pointer transition duration-200 ease-in bg-teal-400"
                />
              </label>
            </div>
            <span
              class="text-gray-700 text-sm font-semibold text-xs text-gray-700"
            >
              カード選択モード
            </span>
          </div>
        </div>
        <table
          class="table-auto w-full"
        >
          <thead>
            <tr>
              <td
                class="px-2 py-1 sticky top-0 bg-white"
              >
                <span
                  class="text-xs font-bold text-gray-600"
                >
                  Round
                </span>
              </td>
            </tr>
          </thead>
          <tbody />
        </table>
        <p
          class="text-red-400"
        />
        <div
          class="text-center mt-4"
        >
          <button
            class="btn bg-gradient-to-r from-teal-400 to-blue-500 text-white"
          >
            add Round
          </button>
        </div>
      </section>
      <section
        class="bg-white p-4 rounded-lg shadow-xl"
      >
        <h2
          class="text-2xl font-bold text-gray-800 mb-4"
        >
          Step.3 Results
        </h2>
        <div>
          <table
            class="table-fixed w-full"
          >
            <thead>
              <tr>
                <th
                  class="border sticky top-0 bg-white"
                >
                  <span
                    class="text-xs font-bold text-gray-500"
                  >
                    Round
                  </span>
                </th>
              </tr>
            </thead>
            <tbody />
            <tfoot>
              <tr
                class="bg-gray-100"
              >
                <th
                  class="border"
                >
                  <span
                    class="text-xs font-bold text-gray-700"
                  >
                    合計
                  </span>
                </th>
              </tr>
              <tr>
                <th
                  class="border"
                >
                  <span
                    class="text-xs font-bold text-gray-700"
                  >
                    合計(x1)
                  </span>
                </th>
              </tr>
            </tfoot>
          </table>
          <div
            class="mt-4"
          >
            <div
              class="recharts-wrapper"
              style="position: relative; cursor: default; width: 400px; height: 400px;"
            >
              <svg
                class="recharts-surface"
                height="400"
                version="1.1"
                viewBox="0 0 400 400"
                width="400"
              >
                <defs>
                  <clippath
                    id="recharts1-clip"
                  >
                    <rect
                      height="360"
                      width="310"
                      x="70"
                      y="5"
                    />
                  </clippath>
                </defs>
                <g
                  class="recharts-cartesian-grid"
                >
                  <g
                    class="recharts-cartesian-grid-horizontal"
                  >
                    <line
                      fill="none"
                      height="360"
                      offset="[object Object]"
                      stroke="#ccc"
                      stroke-dasharray="5 5"
                      width="310"
                      x="70"
                      x1="70"
                      x2="380"
                      y="5"
                      y1="5"
                      y2="5"
                    />
                    <line
                      fill="none"
                      height="360"
                      offset="[object Object]"
                      stroke="#ccc"
                      stroke-dasharray="5 5"
                      width="310"
                      x="70"
                      x1="70"
                      x2="380"
                      y="5"
                      y1="365"
                      y2="365"
                    />
                  </g>
                  <g
                    class="recharts-cartesian-grid-vertical"
                  >
                    <line
                      fill="none"
                      height="360"
                      offset="[object Object]"
                      stroke="#ccc"
                      stroke-dasharray="5 5"
                      width="310"
                      x="70"
                      x1="70"
                      x2="70"
                      y="5"
                      y1="5"
                      y2="365"
                    />
                    <line
                      fill="none"
                      height="360"
                      offset="[object Object]"
                      stroke="#ccc"
                      stroke-dasharray="5 5"
                      width="310"
                      x="70"
                      x1="380"
                      x2="380"
                      y="5"
                      y1="5"
                      y2="365"
                    />
                  </g>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
                >
                  <line
                    class="recharts-cartesian-axis-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    type="category"
                    width="310"
                    x="70"
                    x1="70"
                    x2="380"
                    y="365"
                    y1="365"
                    y2="365"
                  />
                  <g
                    class="recharts-cartesian-axis-ticks"
                  >
                    <g
                      class="recharts-layer recharts-cartesian-axis-tick"
                    >
                      <line
                        class="recharts-cartesian-axis-tick-line"
                        fill="none"
                        height="30"
                        orientation="bottom"
                        stroke="#666"
                        type="category"
                        width="310"
                        x="70"
                        x1="70"
                        x2="70"
                        y="365"
                        y1="371"
                        y2="365"
                      />
                      <text
                        class="recharts-text recharts-cartesian-axis-tick-value"
                        fill="#666"
                        height="30"
                        orientation="bottom"
                        stroke="none"
                        text-anchor="middle"
                        type="category"
                        width="310"
                        x="70"
                        y="373"
                      >
                        <tspan
                          dy="0.71em"
                          x="70"
                        >
                          0
                        </tspan>
                      </text>
                    </g>
                    <g
                      class="recharts-layer recharts-cartesian-axis-tick"
                    >
                      <line
                        class="recharts-cartesian-axis-tick-line"
                        fill="none"
                        height="30"
                        orientation="bottom"
                        stroke="#666"
                        type="category"
                        width="310"
                        x="70"
                        x1="380"
                        x2="380"
                        y="365"
                        y1="371"
                        y2="365"
                      />
                      <text
                        class="recharts-text recharts-cartesian-axis-tick-value"
                        fill="#666"
                        height="30"
                        orientation="bottom"
                        stroke="none"
                        text-anchor="middle"
                        type="category"
                        width="310"
                        x="380"
                        y="373"
                      >
                        <tspan
                          dy="0.71em"
                          x="380"
                        >
                          auto
                        </tspan>
                      </text>
                    </g>
                  </g>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
                >
                  <line
                    class="recharts-cartesian-axis-line"
                    fill="none"
                    height="360"
                    orientation="left"
                    stroke="#666"
                    type="number"
                    width="60"
                    x="10"
                    x1="70"
                    x2="70"
                    y="5"
                    y1="5"
                    y2="365"
                  />
                  <g
                    class="recharts-cartesian-axis-ticks"
                  />
                </g>
              </svg>
              <div
                class="recharts-tooltip-wrapper"
                style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
              >
                <div
                  class="recharts-default-tooltip"
                  style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
                >
                  <p
                    class="recharts-tooltip-label"
                    style="margin: 0px;"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </section>
      <section
        class="bg-white p-4 rounded-lg shadow-xl"
      >
        <h2
          class="text-2xl font-bold text-gray-800 mb-4"
        >
          Step.4 Clear
        </h2>
        <div
          class="text-center mt-4"
        >
          <button
            class="btn bg-gray-800 text-white"
          >
            リセット
          </button>
        </div>
      </section>
    </main>
    <footer
      class="p-4 border-t flex justify-center items-center space-x-4"
    >
      <p
        class="text-center"
      >
        Powered by Ryuta Kojima
      </p>
      <a
        class="bg-black text-white text-sm font-semibold py-1 px-4 rounded-full"
        href="https://github.com/RyutaKojima/UnoScore"
      >
        GitHub
      </a>
    </footer>
  </div>
</DocumentFragment>
`;
